api/envoy/extensions/filters/http/rate_limit_quota/v3/rate_limit_quota.proto

```diff
@@ -0,0 +1,159 @@
+      // [#comment:TODO(sergiitk): what to do after that?]
+      google.protobuf.Duration time_to_live = 1;
+    }
+
+    oneof behavior {
+      option (validate.required) = true;
+
+      // Pre-configured local rate limit strategy.
+      type.v3.RateLimitStrategy fallback_rate_limit = 1;
+
+      // Reuse last quota assignment associated to the bucket.
+      FallbackLastAssignment fallback_last_assignment = 2;
+    }
+  }
+
+  message DenyResponseSettings {
+    // Status code to use to fail requests when quota is exceeded.
+    // [#comment:TODO: figure out the right defaults for these fields!]
+    int32 http_status_code = 1;
+    // For HTTP requests.
+
+    google.rpc.Status deny_grpc_status = 2;
+    // For gRPC requests.
+
+    // A response body to send to the downstream client when the response code
+    // is not OK.
+    bytes http_raw_body = 3;
+
+    // A list of headers to add to the response when the request is throttled.
+    repeated config.core.v3.HeaderValue response_headers_to_add = 4;
+  }
+
+  message BucketIdBuilder {
+    message ValueBuilder {
+      oneof value_specifier {
+        string string_value = 1;
+
+        config.core.v3.TypedExtensionConfig custom_value_builder = 2;
+      }
+    }
+
+    map<string, ValueBuilder> bucket_id_builder = 1;
+  }
+
+  // Contains the instructions for building bucket_id.
+  BucketIdBuilder bucket_id_builder = 1 [(validate.rules).message = {required: true}];
+
+  // The interval at which the data plane is to report quota usage for
+  // this bucket.
+  // When the first request is matched to a bucket without an active
+  // assignment, the data plane immediately sends RateLimitQuotaUsageReports
+  // to the server, signaling it's subscribed for the quota assignments
+  // in that bucket. After that, the data plane is to submit the report
+  // with the given reporting internal.
+  // [#comment:nanos to seconds = 0.001 seconds]
+  google.protobuf.Duration reporting_interval = 2 [(validate.rules).duration = {
+    required: true
+    gt {nanos: 1000000}
+  }];
+
+  // [#comment:TODO(sergiitk): behavior on unset]
+  DenyResponseSettings deny_response_settings = 3;
+
+  // How the data plane should treat requests matched into this bucket
+  // while waiting for the first assignment returned by the RLQS server.
+  // When unset, the default behavior is to allow all requests.
+  NoAssignmentBehavior no_assignment_behavior = 4;
+
+  // Specifies a behavior when the bucket's assignment has expired, and
+  // hasn't been refresh due to server's unavailability.
+  FallbackBehavior fallback_behavior = 5;
+
+  // A duration after which, if no requests are matched to the bucket,
+  // the bucket is considered inactive. The client is to send a message to
+  // the server indicating it's no longer tracking the bucket.
+  google.protobuf.Duration bucket_inactivity_timeout = 6
+      [(validate.rules).duration = {required: true}];
+}
+
+message RateLimitQuotaPerRoute {
```

Comment:
Do we need a different message type for per-route configuration?  What do we want to allow to be overridden on a per-route basis?

Note that we could just use the same proto message for both the top-level config and the per-route config.

Reply:
I haven't done many of these, and will need help with per-route filter config settings. I thought we were going with unifying OverrideOptions, like we discussed here: https://github.com/envoyproxy/envoy/pull/18044/files#r714861992

But it'll be great if the message can be the same.

Reply:
@sergiitk  I feel the issue you mentioned is relevant but not exactly same. We can use this thread to refresh the discussion for that issue.

My understanding:

- Quick recap of the issue in [#18044](https://github.com/envoyproxy/envoy/pull/18044) and [#18118](https://github.com/envoyproxy/envoy/issues/18118) :
The problem there is that [`message RateLimit`](https://github.com/envoyproxy/envoy/blob/4e486e1d336fd0e67ea4f1ee27475daaf6291321/api/envoy/config/route/v3/route_components.proto#L1684) is defined in route_componment.proto and referenced in vhost/route/clusterWighet levels, which is bad because they are tightly tied.
The direction we want to go is that switching to use new `message RateLimitConfig` in rate_limit.proto added in [#18044](https://github.com/envoyproxy/envoy/pull/18044) and referencing it via `typed_per_filter_config `that is already available in those 3 vhost/route/clusterWighet levels.

- The `perRouteConfig`(i.e., RateLimitQuotaPerRoute) you are defining here is the second template argument for creating HTTP filter extension. By default, it is equals to `FilterConfig`  See [factoryBase](https://github.com/envoyproxy/envoy/blob/4e486e1d336fd0e67ea4f1ee27475daaf6291321/source/extensions/filters/http/common/factory_base.h#L14)

- So if there is something need to be overridden on more specific levels (e.g. `route`), the `perRouteConfig` here could be the message struct just has `OverrideOptions ` to define its overridden behavior, and the actual overridden config content is still provided through the typed_per_filter_config from corresponding config level mentioned above.
  If there is nothing need to be overridden, we just omiT this config and use default top level one.

Does this make sense to you guys?

Reply:
I don't think we want to reuse any of the configs for the existing RLS filter, because we're building something new here.  I think we should just structure this in a way that makes sense for the new filter without regard for how the old one worked.

Once we do that, we might even be able to just say that we're addressing #18118 by deprecating the old RLS filter in favor of the RLQS filter. :)  Or we can do the work described in that issue and continue to support both.  But either way, I think the configs of the two filters are orthogonal to each other.

In the case of this new filter, I was originally thinking that we'd just want to be able to override any of the three fields here, but I now think that we probably don't want to be able to override the RLQS server or domain, only the matching tree.  If we need a different RLQS server or domain, we'd probably need to configure a second instance of the RLQS filter.  So I suggest having a separate `RateLimitQuotaOverrideConfig` message that includes only a duplicate of the `bucket_matchers` field from the top-level config.

Reply:
Yea. I was refreshing the discussion for [#18118 ](https://github.com/envoyproxy/envoy/issues/18118) and not saying we should reuse the config:) I feel the deprecating the old RLS is up to discussion and will be a long-term effort if we do.

For this new design, what is your vision for it interacting with local rate limiting mechanism? Maybe we can still use this existing [local rate limiting](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_features/local_rate_limiting#arch-overview-local-rate-limit). This allows rate limiting without calling other service, which is helpful for the scenario, for example, RLQS server is down.

Reply:
The `RateLimitQuotaBucketSettings` should be part of the `RateLimitQuotaFilterConfig` and will specify the rate limit buckets applicable to all requests. Then in addition the `RateLimitQuotaBucketSettings` can be specified at the vhost and route level to provide bucket configuration for specific vhosts and routes. There should be rules that define how bucket configuration is merged.

Reply:
I don't think it's feasible to merge the rules like that, because there's no general algorithm for merging a complete matching tree.  I believe that the `RateLimitQuotaBucketSettings` proto is going to be plugged into the nodes of the matching tree rather than being a top-level field in the config.

I guess we could do something like provide another custom match action that says "execute the match tree from a higher level", which we could then plugin to the match tree in the override config.  But that assumes that any new matching rules in the override config want to be the root of the match tree instead of plugging into a particular node in a sub-tree, so it doesn't really address the general case.  And I think it would wind up being harder for people to understand.

In the general case, I think it may be simpler to just have the control plane override the entire matching tree for each override config.

Reply:
Sorry, the only member that makes sense here is `xds.type.matcher.v3.Matcher bucket_matchers;`, such that more specific overrides use the same RLQS service as the top level config.

Reply:
Addressed in https://github.com/envoyproxy/envoy/pull/19793/commits/d6a23241f24983b997e043bf0bbd032f5821230f.